File: Java/1561_Maximum_Number_of_Coins_You_Can_Get.java
class Solution {
    int[] sortedArray;
    public int maxCoins(int[] piles) {
        int result = 0;
        sortedArray = piles;
        
        Arrays.sort(sortedArray);
        for(int i = sortedArray.length - 2; i >= sortedArray.length/3 ; i-=2){
            result += sortedArray[i];
            //System.out.println(sortedArray[i]);
        }
        
        /*
        //quick sort
        quicksort(0, piles.length-1);
        
        for(int i = 1; i < sortedArray.length*2/3; i+=2){
            result += sortedArray[i];
            //System.out.println(sortedArray[i]);
        }
        */
        return result;
    }
    /*
    private void quicksort(int left, int right){
        if(left >= right){
            return;
        }
        int key = sortedArray[left];
        int i = left;
        int j = right;
        while(i != j){
            while(sortedArray[j] < key  && i<j){
                j--;
            }
            while(sortedArray[i] >= key && i<j){
                i++;
            }
            if(j > i){
                int temp = sortedArray[i];
                sortedArray[i] = sortedArray[j];
                sortedArray[j] = temp;
            }
        }
        sortedArray[left] = sortedArray[i];
        sortedArray[i] = key;
        
        quicksort(left, i-1);
        quicksort(i+1, right);
    }
    */
}

File: CS/1561_Maximum_Number_of_Coins_You_Can_Get.cs
public class Solution {
    public int MaxCoins(int[] piles) {
        int result = 0;
        //Array.Sort(piles);
        quicksort(ref piles, 0, piles.Length-1);
        //bubble sort
        /*
        for(int i = 0; i < piles.Length; i++){
            for(int j = 0; j < piles.Length-i-1; j++){
                if(piles[j] > piles[j+1]){
                    int temp = piles[j];
                    piles[j] = piles[j+1];
                    piles[j+1] = temp;
                }
            }
        }
        */
        
        int n = 0;
        for(int i = piles.Length-2; i>=0 && n < piles.Length; i-=2){
            result += piles[i];
            n += 3;
        }
        return result;
    }
    private void quicksort(ref int[] arr, int l, int r){
        if(l > r){
            return;
        }
        int key = arr[l];
        int i = l;
        int j = r;
        while(i != j){
            while(arr[j] > key && i < j){
                j--;
            }
            while(arr[i] <= key && i < j){
                i++;
            }
            if(i < j){
                int t = arr[i];
                arr[i] = arr[j];
                arr[j] = t;
            }
        }
        arr[l] = arr[i];
        arr[i] = key;
        
        quicksort(ref arr, l, i-1);
        quicksort(ref arr, i+1, r);
    }
}
